qemu/tests/hugepage_specify_node.py

```diff
@@ -0,0 +1,66 @@
+import logging
+
+from avocado.utils import memory
+
+from virttest import env_process
+from virttest import error_context
+from virttest import utils_misc
+from virttest import utils_numeric
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Qemu allocate hugepage from node.
+    Steps:
+    1) System setup hugepage on host.
+    2) Mount this hugepage to /mnt/kvm_hugepage.
+    3) Boot guest only allocate hugepage from node.
+    4) Check the hugepage used only from node.
+    :params test: QEMU test object.
+    :params params: Dictionary with the test parameters.
+    :params env: Dictionary with test environment.
+    """
+    error_context.context("Setup huge pages by default way.", logging.info)
+    hugepage_size = int(memory.get_huge_page_size())
+    mem_size = int(utils_numeric.normalize_data_size(params["mem"], "B"))
+    params["target_hugepages"] = (mem_size // hugepage_size) * 2
+    params["setup_hugepages"] = "yes"
+
+    error_context.context("Get host numa topological structure.", logging.info)
+    host_numa_node = utils_misc.NumaInfo()
+    node_list = host_numa_node.online_nodes
+    pre_hugepages_free = {}
+    for node_id in node_list:
+        error_context.base_context("Check preprocess HugePages Free on host "
+                                   "numa node %s." % node_id, logging.info)
+        node_memfree = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                                 "MemFree"))
+        value = int(host_numa_node.read_from_node_meminfo(node_id,
+                                                          "HugePages_Free"))
+        pre_hugepages_free[node_id] = value
+        if node_memfree < mem_size:
+            node_list.remove(node_id)
+
+    for node_id in node_list:
```

Comment:
I guess we need consider when node_list is empty.

Reply:
yes, thanks for the reminder.